business/pokemon.go

```diff
@@ -31,35 +33,141 @@ func NewPokemonBusiness(
 	log.Println("enter to search and store pokemon by id!!!")
 	pokemonAPI, err := s.serviceAPI.GetPokemonFromAPI(id)
 	if err != nil {
-		return nil, err
+		return model.PokemonDTO{}, err
 	}
-	pokemon, errRepository := s.pokemonRepository.StoreToCSV(*pokemonAPI)
+	pokemon, errRepository := s.pokemonRepository.StoreToCSV(pokemonAPI)
 	if errRepository != nil {
-		return nil, errRepository
+		return model.PokemonDTO{}, errRepository
 	}
 	return pokemon, nil
 }
+
+func (s pokemonBusiness) SearchPokemonSingle(typeSearch string, items int, itemsPerWorker int) ([]model.PokemonDTO, *model.ErrorHandler) {
+	log.Println("enter to search pokemons by odd or even!!!")
+	pokemons, err := s.pokemonRepository.GetAll()
+	if err != nil {
+		return make([]model.PokemonDTO, 0), err
+	}
+	if items > len(pokemons) {
+		return make([]model.PokemonDTO, 0), model.NewWrongParameterLimitValueError("items", 0, len(pokemons))
+	}
+	if items < itemsPerWorker {
+		return make([]model.PokemonDTO, 0), model.NewParameterGreaterThanOtherParameterError("items_per_worker", "items")
+	}
+
+	resultPokemons := make([]model.PokemonDTO, 0)
+	var odd bool
+	if typeSearch == "odd" {
+		odd = true
+	} else {
+		odd = false
+	}
+	for _, pokemon := range pokemons {
+		if odd {
+			if pokemon.ID%2 == 0 {
+				resultPokemons = append(resultPokemons, pokemon)
+			}
+		} else {
+			if pokemon.ID%2 != 0 {
+				resultPokemons = append(resultPokemons, pokemon)
+			}
+		}
+	}
+
+	return resultPokemons, nil
+}
+
+func (s pokemonBusiness) SearchPokemonThread(typeSearch string, items int, itemsPerWorker int) ([]model.PokemonDTO, *model.ErrorHandler) {
+	log.Println("enter to search pokemons by odd or even!!!")
+	pokemons, err := s.pokemonRepository.GetAll()
+	if err != nil {
+		return make([]model.PokemonDTO, 0), err
+	}
+	if items > len(pokemons) {
+		return make([]model.PokemonDTO, 0), model.NewWrongParameterLimitValueError("items", 0, len(pokemons))
+	}
+	if items < itemsPerWorker {
+		return make([]model.PokemonDTO, 0), model.NewParameterGreaterThanOtherParameterError("items_per_worker", "items")
+	}
+
+	numWorkers := int(len(pokemons) / itemsPerWorker)
+	pokemonsChan := make(chan model.PokemonDTO, len(pokemons))
+	resultsChan := make(chan model.WorkerResult, len(pokemons))
+
+	pokemonFinalResult := make([]model.PokemonDTO, 0)
+
+	for w := 0; w < numWorkers; w++ {
+		go worker(w+1, typeSearch, pokemonsChan, resultsChan)
+	}
+
+	isReached := false
+	for !isReached {
+		for _, pokemon := range pokemons {
+			pokemonsChan <- pokemon
+			result := <-resultsChan
+			if result.IsValidResult {
+				pokemonFinalResult = append(pokemonFinalResult, result.Pokemon)
+			}
+			if len(pokemonFinalResult) == items {
+				break
+			}
+		}
+		isReached = true
+	}
+	close(pokemonsChan)
+	return pokemonFinalResult, nil
+}
+
+func worker(id int, typeSearch string, pokemonChan <-chan model.PokemonDTO, result chan<- model.WorkerResult) {
```

Comment:
This worker function is not doing much besides validating if an item should be added to the results or not.

This is where you should distribute the work of fetching the data and validate it.